subprojects/messaging/src/main/java/org/gradle/internal/remote/internal/hub/MessageHubBackedObjectConnection.java

```diff
@@ -58,13 +64,18 @@ public void execute(Throwable throwable) {
 
     @Override
     public void useJavaSerializationForParameters(ClassLoader incomingMessageClassLoader) {
-        methodParamClassLoader = incomingMessageClassLoader;
+        methodParamClassLoaders.add(incomingMessageClassLoader);
     }
 
     public <T> void addIncoming(Class<T> type, final T instance) {
-        if (methodParamClassLoader == null) {
-            methodParamClassLoader = type.getClassLoader();
+        addIncoming(type, instance, type.getClassLoader());
+    }
+
+    public <T> void addIncoming(Class<T> type, final T instance, ClassLoader classLoader) {
```

Comment:
I don't think the `classLoader` parameter is used. We should be able to get rid of this overload of `addIncoming()`